atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/persistentlock/LockStore.java

```diff
@@ -0,0 +1,155 @@
+    }
+
+    public static LockStore create(KeyValueService kvs) {
+        kvs.createTable(AtlasDbConstants.PERSISTED_LOCKS_TABLE, AtlasDbConstants.GENERIC_TABLE_METADATA);
+        LockStore lockStore = new LockStore(kvs);
+        ensurePersistedLocksTableIsPopulated(kvs, lockStore);
+        return lockStore;
+    }
+
+    private static void ensurePersistedLocksTableIsPopulated(KeyValueService kvs, LockStore lockStore) {
+        if (lockStore.allLockEntries().isEmpty()) {
+            populateTable(kvs);
+        }
+    }
+
+    @VisibleForTesting
+    void populateTable() {
+        populateTable(keyValueService);
+    }
+
+    private static void populateTable(KeyValueService kvs) {
+        CheckAndSetRequest request = CheckAndSetRequest.newCell(
+                AtlasDbConstants.PERSISTED_LOCKS_TABLE,
+                LOCK_OPEN.cell(),
+                LOCK_OPEN.value());
+        try {
+            kvs.checkAndSet(request);
+        } catch (CheckAndSetException e) {
+            // This can happen if multiple LockStores are started at once. We don't actually mind.
+            // All we care about is that we're in the state machine of "LOCK_OPEN"/"LOCK_TAKEN".
+            // It still might be interesting, so we'll log it.
+            List<String> values = e.getActualValues().stream()
+                    .map(v -> new String(v, StandardCharsets.UTF_8))
+                    .collect(Collectors.toList());
+            log.warn("Encountered a CheckAndSetException when creating the LockStore. This means that two "
+                    + "LockStore objects were created near-simultaneously, and is probably not a problem. "
+                    + "For the record, we observed these values: {}", values);
+        }
+    }
+
+    public LockEntry acquireLock(String reason) throws PersistentLockIsTakenException {
+        LockEntry lockEntry = generateUniqueLockEntry(reason);
+        CheckAndSetRequest request = CheckAndSetRequest.singleCell(AtlasDbConstants.PERSISTED_LOCKS_TABLE,
+                lockEntry.cell(),
+                LOCK_OPEN.value(),
+                lockEntry.value());
+
+        try {
+            keyValueService.checkAndSet(request);
+        } catch (CheckAndSetException e) {
+            Set<LockEntry> heldLocks = allLockEntries();
+            throw new PersistentLockIsTakenException(lockEntry, heldLocks, e);
+        }
+
+        return lockEntry;
+    }
+
+    public void releaseLock(LockEntry lockEntry) throws PersistentLockIsTakenException {
+        CheckAndSetRequest request = CheckAndSetRequest.singleCell(AtlasDbConstants.PERSISTED_LOCKS_TABLE,
+                lockEntry.cell(),
+                lockEntry.value(),
+                LOCK_OPEN.value());
+
+        try {
+            keyValueService.checkAndSet(request);
+        } catch (CheckAndSetException e) {
+            Set<LockEntry> heldLocks = allLockEntries();
+            throw new PersistentLockIsTakenException(lockEntry, heldLocks, e);
```

Comment:
Seems like we may want to leverage that very useful CheckAndSetException here instead of relegating it to just another Throwable in PersistentLockIsTakenException (ie. maintain the typing so that people can pull out the failed check and set values or integrate them into the error message in a more cohesive manner).  Thinking in particular about the cases were allLockEntries differs from the checkAndSet failure because they are separate actions.